### PR TITLE
BWC test workflow

### PR DESCRIPTION
**Pull request by kavilla**
### Description
Adding backwards compatability workflow for OpenSearch and OpenSearch
Dashboards.

Signed-off-by: Kawika Avilla <kavilla414@gmail.com>
 
### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/705
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kavilla** [comment]
Example test results:
```
[ unzip OpenSearch and OpenSearch Dashboards ]
[ Setup the cypress test environment ]

> cypress@5.6.0 postinstall /tmp/tmpfgsde90n/OpenSearch-Dashboards/bwc_tmp/test/node_modules/cypress
> node index.js --exec install


Cypress 5.6.0 is installed in /home/ubuntu/.cache/Cypress/5.6.0

added 222 packages from 152 contributors and audited 222 packages in 3.195s

15 packages are looking for funding
  run `npm fund` for details

found 3 moderate severity vulnerabilities
  run `npm audit fix` to fix them, or `npm audit` for details
Cypress is ready!
[ Setting up the OpenSearch environment for osd-1.1.0 ]
Data has been uploaded and ready to test
Checking the status OpenSearch...
OpenSearch is up!
Checking the OpenSearch Dashboards...
OpenSearch Dashboards is up!
[ Run the backwards compatibility tests for osd-1.1.0 ]

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:      5.6.0                                                                            │
  │ Browser:      Electron 85 (headless)                                                           │
  │ Specs:        1 found (without-security/check_loaded_data.js)                                  │
  │ Searched:     cypress/integration/without-security/check_loaded_data.js                        │
  │ Experiments:  experimentalNetworkStubbing=true                                                 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  without-security/check_loaded_data.js                                           (1 of 1)


  check previously loaded data
    ✓ previous loaded data should exist in dashboards (2465ms)
    Global Flight Dashboard should function properly
      ✓ Global Flight Dashboard is loaded when clicked (4762ms)
      ✓ If set filter for carrier, [Flights] Airline Carrier should show correct carrier (12069ms)
      ✓ If set filter for FlightDelayType, [Flights] Delay Type should filter out selected type (9367ms)
    eCommerce Revenue Dashboard should function properly
      ✓ eCommerce Revenue Dashboard is loaded when clicked (4820ms)
      ✓ If set filter for gender, [eCommerce] Sales by Gender should show one gender (7244ms)
      ✓ If filter out Women's Clothing, [eCommerce] Sales by Category should not show this category (7210ms)


  7 passing (49s)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        7                                                                                │
  │ Passing:      7                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     49 seconds                                                                       │
  │ Spec Ran:     without-security/check_loaded_data.js                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Video)

  -  Started processing:  Compressing to 32 CRF                                                     
  -  Finished processing: /tmp/tmpfgsde90n/OpenSearch-Dashboards/bwc_tmp/test/cypress     (1 second)
                          /videos/without-security/check_loaded_data.js.mp4                         


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  without-security/check_loaded_data.      00:49        7        7        -        -        - │
  │    js                                                                                          │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        00:49        7        7        -        -        -  


====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:      5.6.0                                                                            │
  │ Browser:      Electron 85 (headless)                                                           │
  │ Specs:        1 found (without-security/check_timeline.js)                                     │
  │ Searched:     cypress/integration/without-security/check_timeline.js                           │
  │ Experiments:  experimentalNetworkStubbing=true                                                 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  without-security/check_timeline.js                                              (1 of 1)


  check timeline visualization
    ✓ timeline visualizations should be saved and named correctly (3497ms)
    timeline visualizations should work properly
      ✓ .es(*, kibana1=true) should report search error (11156ms)
      ✓ .es(*, kibana=true) should not report search error (10659ms)
      ✓ .es(*, opensearchDashboards=true) should not report search error (11044ms)
      ✓ .elasticsearch(*, kibana1=true) should report search error (10912ms)
      ✓ .elasticsearch(*, kibana=true) should not report search error (10915ms)
      ✓ .elasticsearch(*, opensearchDashboards=true) should not report search error (10929ms)
      ✓ .opensearch(*, kibana1=true) should report search error (10906ms)
      ✓ .opensearch(*, kibana=true) should not report search error (11003ms)
      ✓ .opensearch(*, opensearchDashboards=true) should not report search error (10875ms)


  10 passing (2m)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        10                                                                               │
  │ Passing:      10                                                                               │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     1 minute, 42 seconds                                                             │
  │ Spec Ran:     without-security/check_timeline.js                                               │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Video)

  -  Started processing:  Compressing to 32 CRF                                                     
  -  Finished processing: /tmp/tmpfgsde90n/OpenSearch-Dashboards/bwc_tmp/test/cypress    (3 seconds)
                          /videos/without-security/check_timeline.js.mp4                            


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  without-security/check_timeline.js       01:42       10       10        -        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        01:42       10       10        -        -        -  


====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:      5.6.0                                                                            │
  │ Browser:      Electron 85 (headless)                                                           │
  │ Specs:        1 found (without-security/check_advanced_settings.js)                            │
  │ Searched:     cypress/integration/without-security/check_advanced_settings.js                  │
  │ Experiments:  experimentalNetworkStubbing=true                                                 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  without-security/check_advanced_settings.js                                     (1 of 1)


  verify the advanced settings are saved
    ✓ the dark mode is on (2146ms)
    ✓ the Timeline default columns field is set to 4 (910ms)
    ✓ the Timeline Maximum buckets field is set to 4 (692ms)


  3 passing (4s)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        3                                                                                │
  │ Passing:      3                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     3 seconds                                                                        │
  │ Spec Ran:     without-security/check_advanced_settings.js                                      │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Video)

  -  Started processing:  Compressing to 32 CRF                                                     
  -  Finished processing: /tmp/tmpfgsde90n/OpenSearch-Dashboards/bwc_tmp/test/cypress    (0 seconds)
                          /videos/without-security/check_advanced_settings.js.mp4                   


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  without-security/check_advanced_set      00:03        3        3        -        -        - │
  │    tings.js                                                                                    │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        00:03        3        3        -        -        -  


====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:      5.6.0                                                                            │
  │ Browser:      Electron 85 (headless)                                                           │
  │ Specs:        1 found (without-security/check_filter_and_query.js)                             │
  │ Searched:     cypress/integration/without-security/check_filter_and_query.js                   │
  │ Experiments:  experimentalNetworkStubbing=true                                                 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  without-security/check_filter_and_query.js                                      (1 of 1)


  check dashboards filter and query
    osx filter and query should work in [Logs] Web Traffic dashboards
      ✓ osx filter and query should exist and be named correctly (3514ms)
      ✓ osx filter and query should function correctly (5075ms)


  2 passing (9s)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        2                                                                                │
  │ Passing:      2                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     8 seconds                                                                        │
  │ Spec Ran:     without-security/check_filter_and_query.js                                       │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Video)

  -  Started processing:  Compressing to 32 CRF                                                     
  -  Finished processing: /tmp/tmpfgsde90n/OpenSearch-Dashboards/bwc_tmp/test/cypress    (0 seconds)
                          /videos/without-security/check_filter_and_query.js.mp4                    


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  without-security/check_filter_and_q      00:08        2        2        -        -        - │
  │    uery.js                                                                                     │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        00:08        2        2        -        -        -  


====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:      5.6.0                                                                            │
  │ Browser:      Electron 85 (headless)                                                           │
  │ Specs:        1 found (without-security/check_default_page.js)                                 │
  │ Searched:     cypress/integration/without-security/check_default_page.js                       │
  │ Experiments:  experimentalNetworkStubbing=true                                                 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  without-security/check_default_page.js                                          (1 of 1)


  verify default landing page work for bwc
    ✓ the overview page is set as the default landing page (1718ms)


  1 passing (2s)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        1                                                                                │
  │ Passing:      1                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     1 second                                                                         │
  │ Spec Ran:     without-security/check_default_page.js                                           │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Video)

  -  Started processing:  Compressing to 32 CRF                                                     
  -  Finished processing: /tmp/tmpfgsde90n/OpenSearch-Dashboards/bwc_tmp/test/cypress    (0 seconds)
                          /videos/without-security/check_default_page.js.mp4                        


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  without-security/check_default_page      00:01        1        1        -        -        - │
  │    .js                                                                                         │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        00:01        1        1        -        -        -  

Closing the running OpenSearch
Closing any usage on port 5601
```

**Comment by kavilla** [comment]
Related OpenSearch Dashboards PR: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/1190

File: src/test_workflow/bwc_test/bwc_test_suite_opensearch.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        super().__init__(
+            work_dir,
+            component,
+            test_config,
+            test_recorder,
+            manifest
+        )
+
+    def get_cmd(self, script, security, manifest_build_location):
+        return f"{script}"
```
Comment (kavilla) [inline src/test_workflow/bwc_test/bwc_test_suite_opensearch.py:32]: This should be updated but current `bwctest.sh` doesn't accept other arguments: https://github.com/opensearch-project/anomaly-detection/blob/main/bwctest.sh

File: src/test_workflow/bwc_test/bwc_test_runner.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+        self.tests_dir = os.path.join(os.getcwd(), "test-results")
+        os.makedirs(self.tests_dir, exist_ok=True)
+        self.test_recorder = TestRecorder(self.args.test_run_id, "bwc-test", self.tests_dir)
+
+    def run(self):
+        with TemporaryDirectory(keep=self.args.keep, chdir=True) as work_dir:
+            all_results = TestSuiteResults()
+            for component in self.components.select(focus=self.args.component):
+                if component.name in self.test_manifest.components:
```
Comment (dblock) [inline src/test_workflow/bwc_test/bwc_test_runner.py:29]: Not for this PR, but we do a similar workflow in many places and I don't love the double-select here, first pass `--component name` to pick a component, but then we also only want components that have a test configuration.

One option would be to fail if `self.args.component not in self.test_manifest.components` with an explicit error, and refactor `components.select` to take an array of components, i.e. `self.components.select(focus=self.test_manifest.components)`.

Another, possibly better, could be to do `self.components.select(focus=self.args.component, configs=self.test_manifest.components, key: "bwc_test")` that would return a set of tuples of component + configuration. The code then becomes quite clean:

```python
for component, config in self.components.select(focus=self.args.component, configs: self.test_manifest.components, key: "bwc_test"):
   test_suite = ...
   ...
```

File: src/test_workflow/bwc_test/bwc_test_suite_opensearch.py (hunk 1 above)
Comment (dblock) [inline src/test_workflow/bwc_test/bwc_test_suite_opensearch.py:32]: `returbn script` for now without `f`

**Review by dblock — APPROVED** [review]
Good work!

Update READMEs wrt bwc tests soon.

File: src/test_workflow/bwc_test/bwc_test_runner.py (hunk 2 above)
Comment (kavilla) [reply src/test_workflow/bwc_test/bwc_test_runner.py:29]: Created a cleanup issue here: https://github.com/opensearch-project/opensearch-build/issues/1604 and assigned myself.

**Review by dblock — APPROVED** [review]

File: manifests/2.0.0/opensearch-dashboards-2.0.0-test.yml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  - name: OpenSearch-Dashboards
+    bwc-test:
+      test-configs:
+        - without-security
+  - name: functionalTestDashboards
+    integ-test:
+      test-configs:
+        - with-security
+        - without-security
+schema-version: '1.0'
```
Comment (peternied) [inline manifests/2.0.0/opensearch-dashboards-2.0.0-test.yml:13]: Nit; Can we move this to the top?

File: src/test_workflow/README.md [hunk 4]
```diff
@@ -76,6 +76,27 @@ Usage:
...
+```bash
+./build.sh manifests/1.3.0/opensearch-1.3.0.yml
+./assemble.sh builds/opensearch/manifest.yml
+./test.sh bwc-test manifests/1.3.0/opensearch-1.3.0-test.yml . # looks for "./builds/opensearch/manifest.yml" and "./dist/opensearch/manifest.yml"
+```
+
+Or run BWC tests against an existing build.
+
+```bash
+./test.sh bwc-test manifests/1.3.0/opensearch-1.3.0-test.yml --paths opensearch=https://ci.opensearch.org/ci/dbc/bundle-build/1.2.0/869/linux/x64 # looks for https://.../builds/opensearch/manifest.yml and https://.../dist/opensearch/manifest.yml
```
Comment (peternied) [inline src/test_workflow/README.md:90]: Thanks for the great documentation

File: src/test_workflow/bwc_test/bwc_test_suite_opensearch.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (peternied) [inline src/test_workflow/bwc_test/bwc_test_suite_opensearch.py:1]: Nit: this class is a whole lot of stuff for effectively two lines of differentiated behavior.  I don't think there is an easier way but food for thought

File: tests/tests_test_workflow/test_bwc_workflow/bwc_test/test_bwc_test_runner_opensearch.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        self.test_manifest.components = {"sql": mock_test_config}
+
+        mock_components = MagicMock()
+        mock_component = MagicMock()
+        mock_component.name = "sql"
+        mock_components.select.return_value = [mock_component]
+        mock_bundle_manifest = MagicMock()
+        mock_bundle_manifest.components = mock_components
+
+        mock_build_manifest = MagicMock()
```
Comment (peternied) [inline tests/tests_test_workflow/test_bwc_workflow/bwc_test/test_bwc_test_runner_opensearch.py:38]: Nit; Looks like you setup mock manifests a couple of times over, might be worthwhile to add MockBuildManifest/MockTestManifest that could condense these unit tests, or at least let them reuse the same setup logic

Comment (peternied) [reply tests/tests_test_workflow/test_bwc_workflow/bwc_test/test_bwc_test_runner_opensearch.py:38]: Note; even if they used a 'canned' construction pattern and then you modified them afterward to fit one test or another that could be useful

**Review by peternied — APPROVED** [review]
Great contribution!

We've got an outstanding item https://github.com/opensearch-project/opensearch-build/issues/1512 to make sure the test workflows have typing information - do you think you could see about adding in typing info in your new classes?  If you'd like to pickup that issue separately that works for me too.

**Comment by kavilla** [comment]
> Great contribution!
> 
> We've got an outstanding item #1512 to make sure the test workflows have typing information - do you think you could see about adding in typing info in your new classes? If you'd like to pickup that issue separately that works for me too.

Not sure what you quiet mean here. Do you have an example? If it requires updates, then probably will need to do this in the integ tests as well. Which might be beneficial to create another clean up task that I can do to get the bwc tests and integ tests aligned.

**Comment by peternied** [comment]
I've merged your change, checkout issue https://github.com/opensearch-project/opensearch-build/issues/1237 for a detailed explanation with some example PRs

**Comment by kavilla** [comment]
> I've merged your change, checkout issue #1237 for a detailed explanation with some example PRs

Will do and will handle some of the clean ups you suggested

**Pull request merged**